### ISSUE TITLE
Make it possible to select which commands are available

### ISSUE THREAD

**Issue by andrerpena** [andrerpena/react-mde#58]
This was inspired by @daanishnasir's issue here: https://github.com/andrerpena/react-mde/issues/57

**Comment by andrerpena** [comment]
I think the easiest solution is just to export every command from here: https://github.com/andrerpena/react-mde/blob/master/src/ReactMdeCommands.tsx

... so you can import just the ones you like, build your own array and pass it as a prop to `react-mde`

**Comment by daanishnasir** [comment]
ah almost! 

It does get rid of any commands that I take out of the toolbar. However the array of commands is dependent on ./ReactMdeCommandHelper (and maybe more files). I'll try making another file importing the helper functions from ReactMdeCommandHelper into ReactMdeCommands and will update this post if it works. 

Example error right now:

`Uncaught ReferenceError: makeACommandThatInsertsBeforeAndAfter is not defined. `

Code: 

```
import * as React from 'react';

// import {
//     // text insertion
//     insertText,
//     insertBefore,
//     insertAfter,
//     insertBreaksBeforeSoThatThereIsAnEmptyLineBefore,
//     insertBreaksAfterSoThatThereIsAnEmptyLineAfter,
//     // others
//     selectCurrentWordIfCaretIsInsideOne,
// } from './ReactMdeTextHelper';

// import {
//     makeList,
//     makeHeader,
//     makeACommandThatInsertsBeforeAndAfter,
// } from './ReactMdeCommandHelper';
// import { TextSelection } from './types/TextSelection';
// import { Command } from './types/Command';
// import { CommandSet } from './types/CommandSet';

const makeHeaderCommand: CommandSet = {
    type: 'dropdown',
    icon: 'header',
    subCommands: [
        {
            content: <p className="header-1">Header</p>,
            execute: (text: string, selection: TextSelection) => {
                return makeHeader(text, selection, '# ');
            },
        },
        {
            content: <p className="header-2">Header</p>,
            execute(text: string, selection: TextSelection) {
                return makeHeader(text, selection, '## ');
            },
        },
        {
            content: <p className="header-3">Header</p>,
            execute(text: string, selection: TextSelection) {
                return makeHeader(text, selection, '### ');
            },
        },
    ],
};

const makeBoldCommand: Command = {
    icon: 'bold',
    tooltip:
        'Add bold text',
    execute:
        (text: string, selection: TextSelection) => {
            return makeACommandThatInsertsBeforeAndAfter(text, selection, '**');
        },
};

const makeItalicCommand: Command = {
    icon: 'italic',
    tooltip:
        'Add italic text',
    execute:
        (text: string, selection: TextSelection) => {
            return makeACommandThatInsertsBeforeAndAfter(text, selection, '_');
        },
};

const makeLinkCommand = {
    icon: 'link',
    tooltip:
        'Insert a link',
    execute:
        (text: string, selection: TextSelection) => {
            const {newText, insertionLength} = insertText(text, '[', selection.start);
            const finalText = insertText(newText, '](url)', selection.end + insertionLength).newText;
            return {
                text: finalText,
                selection: {
                    start: selection.start + insertionLength,
                    end: selection.end + insertionLength,
                },
            };
        },
};

const makeQuoteCommand = {
    icon: 'quote-right',
    tooltip: 'Insert a quote',
    execute:
        (text: string, selection: TextSelection) => {
            selection = selectCurrentWordIfCaretIsInsideOne(text, selection);

            let textInsertion;

            textInsertion = insertBreaksBeforeSoThatThereIsAnEmptyLineBefore(text, selection);
            text = textInsertion.newText;
            selection = textInsertion.newSelection;

            textInsertion = insertBefore(text, '> ', selection);
            text = textInsertion.newText;
            selection = textInsertion.newSelection;

            textInsertion = insertBreaksAfterSoThatThereIsAnEmptyLineAfter(text, selection);
            text = textInsertion.newText;
            selection = textInsertion.newSelection;

            return {
                text,
                selection,
            };
        },
};

// const makeCodeCommand: Command = {
//     icon: 'code',
//     tooltip: 'Insert code',
//     execute:
//         (text = '', selection: TextSelection) => {
//             selection = selectCurrentWordIfCaretIsInsideOne(text, selection);

//             if (text.slice(selection.start, selection.end).indexOf('\n') === -1) {
//                 // when there's no breaking line
//                 return makeACommandThatInsertsBeforeAndAfter(text, selection, '`');
//             }
//             let textInsertion;

//             // insert breaks before, if needed
//             textInsertion = insertBreaksBeforeSoThatThereIsAnEmptyLineBefore(text, selection);
//             text = textInsertion.newText;
//             selection = textInsertion.newSelection;

//             // inserts ```\n before
//             textInsertion = insertBefore(text, '```\n', selection, false);
//             text = textInsertion.newText;
//             selection = textInsertion.newSelection;

//             // inserts ```\n after
//             textInsertion = insertAfter(text, '\n```', selection);
//             text = textInsertion.newText;
//             selection = textInsertion.newSelection;

//             // insert breaks after, if needed
//             textInsertion = insertBreaksAfterSoThatThereIsAnEmptyLineAfter(text, selection);
//             text = textInsertion.newText;
//             selection = textInsertion.newSelection;

//             return {text, selection};
//         },
// };

const makeImageCommand: Command = {
    icon: 'picture-o',
    tooltip: 'Insert a picture',
    execute:
        (text: string, selection: TextSelection) => {
            const {newText, insertionLength} = insertText(text, '![', selection.start);
            const finalText = insertText(newText, '](image-url)', selection.end + insertionLength).newText;
            return {
                text: finalText,
                selection: {
                    start: selection.start + insertionLength,
                    end: selection.end + insertionLength,
                },
            };
        },
};

const makeUnorderedListCommand: Command = {
    icon: 'list-ul',
    tooltip: 'Add a bulleted list',
    execute:
        (text: string, selection: TextSelection) => {
            return makeList(text, selection, '- ');
        },
};

const makeOrderedListCommand: Command = {
    icon: 'list-ol',
    tooltip: 'Add a numbered list',
    execute:
        (text: string, selection: TextSelection) => {
            return makeList(text, selection, (item: string, index: number) => `${index + 1}. `);
        },
};

export const defaultCommands = [
    [makeHeaderCommand, makeBoldCommand, makeItalicCommand],
    [makeLinkCommand, makeQuoteCommand, makeImageCommand],
    [makeUnorderedListCommand, makeOrderedListCommand],
];

```

Component with ReactMde in case anyone is confused on what we're talking about on how to do this.
```

import {defaultCommands} from "./ReactMdeCommands.jsx";


               <ReactMde
                    textAreaProps={{                  
                        id: 'ta1',
                        name: 'ta1',
                        rows : this.props.rows,
                        title: this.props.title,
                        onBlur: this.onBlur.bind(this),
                        readOnly: this.props.readOnly
                        
                    }}
                    value={this.state.reactMdeValue}
                    onChange={this.handleValueChange}
                    commands={defaultCommands}
                />
            </div>
        );

```

**Comment by daanishnasir** [comment]
It works fine from what i can see. I have removed the commands I don't need now. Just need ALL the files that ReactMdeCommands imports from (ReacMdeCommandHelper, ReactMdeTextHelper, all the type files needed).  Luckily typescript files automatically map to js files so it was pretty easy.  This is a great repo; thanks Andre for being really responsive to the issues. The only thing I can think of to make this even better would be the **_option_** to have a preview command where it toggles to preview instead of having the preview hang below the text.  Something like this http://jaszhix.com/react-markdown-editor-hybrid/

**Comment by andrerpena** [comment]
Sorry for the late response. That's very good to know that you got it working. Thank you for the kind words.

From your comments, I understand that you were able to make to work because of the "allowJs" option of the TypeScript compiler. The ideal solution would be to make all the helper functions, like the ones here: 
https://github.com/andrerpena/react-mde/blob/master/src/ReactMdeTextHelper.ts importable through the index.

What I mean is, if you do something like this:

    import * as ReactMde from 'react-mde';

Now `ReactMde` should allow you access to the helper methods to, otherwise it is difficult for you to create your own commands.

So, to summarize, this is what I'm going to do to close this issue:

- In addition to exporting the default commands as an array of arrays, as it is today, I'll export each individual command so you can build your own list by creating your own array.
- I'll make sure all helper functions are importable.

Fortunately I can do this without breaking the existing code so I can increase just a minor version. I'll prioritize this.

Thanks for your help.

**Comment by andrerpena** [comment]
Oh, and regarding your suggestion of making the preview "toggable", I think you mean like GitHub

![image](https://user-images.githubusercontent.com/1581429/33102631-d40d769a-cf04-11e7-9cbd-96ca06753912.png).

This is a reasonable request. I'll implement it (Feel free to create an issue for it if you like).

I think I'm going to follow this direction for `react-mde`: I'll give the user 2 options:

1. Use the full-blown component, with the toolbar, textarea and preview, with options to disable or toggle the preview behavior.
2. In addition, provide each of the 3 components I mentioned above so you can compose your own, I mention this here: https://github.com/andrerpena/react-mde/issues/51

**Comment by daanishnasir** [comment]
Thanks, this would be perfect. Will actively check on up on this issue. Currently I have too many typescript files in order to pick and choose which commands I want available. I really would like to to have no typescript files at all.

**Comment by andrerpena** [comment]
This is basically completed in https://github.com/andrerpena/react-mde/tree/2.0-rc1 but because of some minor breaking changes in the way you import stuff, I had to increment the major version and I'll only have time to document it this Monday or Tuesday. Then I'll merge it to master.

**Comment by daanishnasir** [comment]
Awesome, I was playing around with it to see if I could get it working with the current version. I'm able to hide the commands by making a new array and splicing what i need from the ReactMdeCommands array. Just a temporary solution.

```

import React, { Component } from 'react';
import ReactMde, { ReactMdeCommands } from 'react-mde';


export class Test extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            reactMdeValue: {text: ""},
        };


    }

    handleValueChange = (value: ReactMdeTypes.Value) => {
        this.setState({reactMdeValue: value});

    }

    render() {
       
```
    let realCommands = [...ReactMdeCommands.slice(0,1), ...ReactMdeCommands.slice(2)]
       
```
        return (
            <div className="container">
             
                <ReactMde
                    textAreaProps={{
                        id: "ta1",
                        name: "ta1",
                    }}
                    value={this.state.reactMdeValue}
                    onChange={this.handleValueChange}
                    commands={realCommands}
                />
            </div>
```

**Comment by andrerpena** [comment]
Yes that would work. Good job. Now you can import each individual command with the 2.* release. I still need to document a bit better but it works.
Thanks for reporting.
